src/context/CartContext.tsx

```diff
@@ -0,0 +1,180 @@
+
+type CartProviderProps = {
+  children: ReactNode;
+};
+
+type CartContext = {
+  state: {
+    isGiftWrapSelected: boolean;
+    cartItems: CartItem[];
+  };
+  actions: {
+    getItemQuantity: (id: number, size: string) => number;
+    increaseCartQuantity: (id: number, size: string) => void;
+    decreaseCartQuantity: (id: number, size: string) => void;
+    removeFromCart: (id: number, size: string) => void;
+    addGiftWrap: () => void;
+    toggleGiftWrap: () => void;
+  };
+  meta: {
+    totalSumWithGiftWrap: number;
+    totalSum: number;
+    cartQuantity: number;
+  };
+};
+
+export const GIFT_WRAP_PRICE = 10;
+
+//CART CONTEXT
+export const CartContext = createContext<CartContext | undefined>(undefined);
+
+export const useShoppingCart = () => {
+  const context = useContext(CartContext);
+  if (!context) {
+    throw new Error('useShoppingCart must be used within a CartProvider');
+  }
+  return context;
+};
+
+// LOCAL STORAGE
+export function useLocalStorage<T>(key: string, initialValue: T | (() => T)) {
+  const [value, setValue] = useState<T>(() => {
+    const jsonValue = localStorage.getItem(key);
+    if (jsonValue !== null && jsonValue !== undefined) {
+      try {
+        return JSON.parse(jsonValue);
+      } catch (error) {
+        console.error('Error parsing JSON from localStorage:', error);
+      }
+    }
+    if (typeof initialValue === 'function') {
+      return (initialValue as () => T)();
+    } else {
+      return initialValue;
+    }
+  });
+
+  useEffect(() => {
+    localStorage.setItem(key, JSON.stringify(value));
+  }, [key, value]);
+
+  return [value, setValue] as [typeof value, typeof setValue];
+}
+
+//CART PROVIDER
+export const CartProvider: React.FC<CartProviderProps> = ({ children }) => {
+  const [cartItems, setCartItems] = useLocalStorage<CartItem[]>(
+    'shoppingCart',
+    [],
+  );
+  const [isGiftWrapSelected, setIsGiftWrapSelected] = useState(false);
+  const { products } = useProducts();
+  const { isOpen } = useCartOverlay();
+
+  function getItemQuantity(id: number, size: string) {
+    const item = cartItems.find(item => item.id === id && item.size === size);
+    return item ? item.quantity : 0;
+  }
+
+  function increaseCartQuantity(id: number, size: string) {
+    setCartItems(currItems => {
+      const newItem = { id, size, quantity: 1 } as CartItem;
+      const existingItemIndex = currItems.findIndex(
+        item => item.id === id && item.size === size,
+      );
+      if (existingItemIndex !== -1) {
+        const updatedItems = [...currItems];
+        updatedItems[existingItemIndex].quantity++;
+        return updatedItems;
+      } else {
+        return [...currItems, newItem];
+      }
+    });
+  }
+
+  function decreaseCartQuantity(id: number, size: string) {
+    setCartItems(currItems => {
+      const existingItem = currItems.find(
+        item => item.id === id && item.size === size,
+      );
+      if (existingItem) {
```

Comment:
Zagnieżdżanie ifów to zwykle zła praktyka.
Można to zbatchować w ten sposób:
`if(existingItem && exisitingItem.quantity === 1)`.
Czasami nie możemy tego zbatchować, ale możemy odwrócić:
`if(!existingItem) return;` - tutaj myślę, że lepiej właśnie to wykorzystać. Nawet nie trzeba wywoływać `setCartItems`.
Ja bym to zrobił tak:

```ts
function decreaseCartQuantity(id: number, size: string) {
 const existingItem = cartItems.find(item => item.id === id && item.size === size);

if(!existingItem) return;
// to co poniżej nigdy się nie wykona jeśli itema nie ma, nie pójdzie też niepotrzebny rerender.
setCartItems(...)
}
```
I już czytelność byłaby lepsza :)